akka-projection-core/src/main/scala/akka/projection/internal/InternalProjectionState.scala

```diff
@@ -202,30 +204,29 @@ private[akka] abstract class InternalProjectionState[Offset, Envelope](
           () => handler.process(envelopes))
       }
 
-      // FIXME create a SourceProvider trait that implies mergeable offsets?
-      if (sourceProvider.isOffsetMergeable) {
-        val batches = envelopesAndOffsets.groupBy {
-          // FIXME matched source provider should always be MergeableOffset
-          case ProjectionContextImpl(offset: MergeableOffset[_, _], _) =>
-            // FIXME we can assume there's only one actual offset per envelope, but there should be a better way to represent this
-            val mergeableKey = offset.entries.head._1.asInstanceOf[MergeableKey]
-            mergeableKey.surrogateKey
-          case _ =>
-            // should never happen
-            throw new IllegalStateException("The offset should always be of type MergeableOffset")
-        }
-
-        // process batches in sequence, but not concurrently, in order to provide singled threaded guarantees
-        // to the user envelope handler
-        serialize(batches) { (surrogateKey, partitionedEnvelopes) =>
-          logger.debug("Processing grouped envelopes for MergeableOffset with key [{}]", surrogateKey)
-          processEnvelopes(partitionedEnvelopes)
-        }
-
-      } else {
-        processEnvelopes(envelopesAndOffsets)
+      sourceProvider match {
+        case _: MergeableOffsetSourceProvider[_, Offset, Envelope] =>
+          val batches = envelopesAndOffsets
+            .asInstanceOf[immutable.Seq[ProjectionContextImpl[MergeableOffset[MergeableKey, _], Envelope]]]
```

Comment:
I'm casting here to access the `ProjectionContextImpl.context` member.  I played with making this grouping code an implementation in `MergeableOffsetSourceProvider`, but it doesn't make sense without making the offset member publicly available in `ProjectionContext`.  I'm not sure it makes sense to add the member just for this use case though.

Reply:
The previous `case ProjectionContextImpl(offset: MergeableOffset[_, _], _) =>` was less ugly, even though it had the throw case. Can't we still use that? maybe as
```
case ProjectionContextImpl(offset: MergeableOffset[MergeableKey, _], _) =>
```
might need an `@unchecked` annotation somewhere to avoid warning

Reply:
I think I like the explicit casting better, but I've changed it as you suggest.  `scalafmt` makes it ugly again.